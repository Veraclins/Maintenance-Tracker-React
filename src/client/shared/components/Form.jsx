import React from 'react';
import { Link } from 'react-router-dom';
import { PropTypes } from 'prop-types';
import DInput from './Input';
import DSelectInput from './Select';
import DTextArea from './TextArea';

export const Form = ({
  inputs,
  handleInputChange,
  handleSelectChange,
  handleTextAreaChange,
  handleSubmit,
  formTitle,
  errors,
  selects,
  textAreas,
  formLinks,
}) => (
  <div className="form_container">
    <div className="white-popup">
      <div className="hbspt-form-title">
        <h3><strong>{formTitle}</strong></h3>
      </div>
      <div className="hbspt-form-container">
        <form
          className="hs-form stacked"
          onSubmit={handleSubmit}
        >
          {Object.entries(inputs).map(([input, value]) => (
            <DInput
              inputLabel={`${input[0].toUpperCase()}${input.slice(1)}`}
              type={value.type}
              name={input}
              required={value.required}
              key={input}
              placeholder={value.placeholder}
              error={errors[input]}
              value={value.value}
              handleChange={handleInputChange}
            />
          ))}
          {Object.entries(selects).map(([select, value]) => (
            <DSelectInput
              inputLabel={`${select[0].toUpperCase()}${select.slice(1)}`}
              name={select}
              required={value.required}
              error={errors[select]}
              placeholder={value.placeholder}
              key={select}
              value={value.value}
              options={value.options}
              handleChange={handleSelectChange}
            />
          ))}
          {Object.entries(textAreas).map(([textArea, value]) => (
            <DTextArea
              inputLabel={`${textArea[0].toUpperCase()}${textArea.slice(1)}`}
              name={textArea}
              required={value.required}
              error={errors[textArea]}
              placeholder={value.placeholder}
              value={value.value}
              key={textArea}
              handleChange={handleTextAreaChange}
            />
          ))}
          <div className="hs-submit">
            <div className="actions">
              <input
                type="submit"
                value="Submit"
                className="hs-button primary large"
              />
            </div>
          </div>
          {formLinks.link
            ? (
              <div className="form_links">
                <p>{formLinks.label} <Link to={formLinks.link}>{formLinks.text}</Link></p>
              </div>
            ) : false}
        </form>
      </div>
    </div>
  </div>
);

Form.propTypes = {
  handleInputChange: PropTypes.func.isRequired,
  handleSelectChange: PropTypes.func,
  handleTextAreaChange: PropTypes.func,
  handleSubmit: PropTypes.func.isRequired,
  formTitle: PropTypes.string.isRequired,
  errors: PropTypes.shape({}).isRequired,
  selects: PropTypes.shape({}),
  textAreas: PropTypes.shape({}),
  inputs: PropTypes.shape({}).isRequired,
  formLinks: PropTypes.shape({}),
};

Form.defaultProps = {
  selects: {},
  textAreas: {},
  formLinks: {},
  handleSelectChange: () => null,
  handleTextAreaChange: () => null,
};
export default Form;
