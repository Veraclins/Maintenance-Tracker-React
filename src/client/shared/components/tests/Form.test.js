import React from 'react';
import { shallow } from 'enzyme';
import DInput from '../Input';

import { Form } from '../Form';
import DSelectInput from '../Select';
import DTextArea from '../TextArea';

const mockFunction = jest.fn();
describe('Test the Form component', () => {
  it('renders without crashing', () => {
    const inputs = {};
    const errors = {};
    const wrapper = shallow(<Form
      inputs={inputs}
      errors={errors}
      handleInputChange={mockFunction}
      handleSubmit={mockFunction}
      formTitle="Default Form"
    />);
    expect(wrapper.find('form')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders all elements when there are inputs', () => {
    const inputs = {
      email: {
        type: 'email',
        value: '',
        placeholder: 'Enter your email',
        required: true,
      },
      password: {
        type: 'password',
        value: '',
        placeholder: 'Enter your password',
        required: true,
      },
    };
    const errors = {};
    const formLinks = {
      label: "Don't have an account?",
      link: '/signup',
      text: 'Create one',
    };
    const wrapper = shallow(<Form
      inputs={inputs}
      errors={errors}
      formLinks={formLinks}
      handleInputChange={mockFunction}
      handleSubmit={mockFunction}
      formTitle="Log into Your Account"
    />);
    expect(wrapper.find('div')).toHaveLength(7);
    expect(wrapper.find('h3')).toHaveLength(1);
    expect(wrapper.find(DInput)).toHaveLength(2);
    expect(wrapper.find('input.hs-button')).toHaveLength(1);
  });

  it('renders all elements when there are select inputs', () => {
    const options = [
      { value: '', label: 'Select Device Type' },
      { value: 'Laptop', label: 'Laptop Pc' },
    ];
    const select = {
      device: {
        options,
        placeholder: 'Select your device type',
        required: true,
        value: '',
      },
    };
    const errors = {};
    const formLinks = {
      label: "Don't have an account?",
      link: '/signup',
      text: 'Create one',
    };
    const inputs = {};
    const wrapper = shallow(<Form
      inputs={inputs}
      selects={select}
      errors={errors}
      formLinks={formLinks}
      handleInputChange={mockFunction}
      handleSelectChange={mockFunction}
      handleSubmit={mockFunction}
      formTitle="Log into Your Account"
    />);
    expect(wrapper.find('div')).toHaveLength(7);
    expect(wrapper.find('h3')).toHaveLength(1);
    expect(wrapper.find(DSelectInput)).toHaveLength(1);
    expect(wrapper.find('input.hs-button')).toHaveLength(1);
  });

  it('renders all elements when there are text areas', () => {
    const textArea = {
      description: {
        value: '',
        required: true,
        placeholder: 'Enter the description of the request',
      },
    };
    const errors = {};
    const formLinks = {
      label: "Don't have an account?",
      link: '/signup',
      text: 'Create one',
    };
    const inputs = {};
    const wrapper = shallow(<Form
      inputs={inputs}
      textAreas={textArea}
      errors={errors}
      formLinks={formLinks}
      handleInputChange={mockFunction}
      handleTextAreaChange={mockFunction}
      handleSubmit={mockFunction}
      formTitle="Log into Your Account"
    />);
    expect(wrapper.find('div')).toHaveLength(7);
    expect(wrapper.find('h3')).toHaveLength(1);
    expect(wrapper.find(DTextArea)).toHaveLength(1);
    expect(wrapper.find('input.hs-button')).toHaveLength(1);
  });
  it('uses default props when not supplied', () => {
    expect(Form.defaultProps.handleSelectChange).toBeDefined();
    expect(Form.defaultProps.handleTextAreaChange).toBeDefined();
    const result = Form.defaultProps.handleTextAreaChange();
    expect(result).toBe(null);
    const select = Form.defaultProps.handleSelectChange();
    expect(select).toBe(null);
  });
});
