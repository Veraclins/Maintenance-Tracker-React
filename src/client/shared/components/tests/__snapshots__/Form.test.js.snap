// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Form component renders without crashing 1`] = `
<div
  className="form_container"
>
  <div
    className="white-popup"
  >
    <div
      className="hbspt-form-title"
    >
      <h3>
        <strong>
          Default Form
        </strong>
      </h3>
    </div>
    <div
      className="hbspt-form-container"
    >
      <form
        className="hs-form stacked"
        onSubmit={[MockFunction]}
      >
        <div
          className="hs-submit"
        >
          <div
            className="actions"
          >
            <input
              className="hs-button primary large"
              type="submit"
              value="Submit"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
